mqtt/mqtt3/src/client/subscriptions.rs

```diff
@@ -223,66 +223,42 @@ impl State {
-                            );
-                        }
-                    }
-
-                    BatchedSubscriptionUpdate::Unsubscribe(unsubscribe_from) => {
-                        for unsubscribe_from in unsubscribe_from {
-                            current_subscriptions.remove(&**unsubscribe_from);
-                        }
-                    }
-                }
-            }
-
-            let mut target_subscriptions = current_subscriptions.clone();
+            let mut target_subscriptions = std::collections::BTreeMap::new();
+            let mut target_unsubscriptions = std::collections::BTreeMap::new();
 
             while let Some(subscription_update) =
                 self.subscription_updates_waiting_to_be_sent.pop_front()
             {
                 match subscription_update {
-                    SubscriptionUpdate::Subscribe(subscribe_to) => target_subscriptions.insert(
-                        std::borrow::Cow::Owned(subscribe_to.topic_filter),
-                        subscribe_to.qos,
-                    ),
+                    SubscriptionUpdate::Subscribe(subscribe_to) => {
+                        target_unsubscriptions.remove(&subscribe_to.topic_filter);
+                        target_subscriptions.insert(
+                            std::borrow::Cow::Owned(subscribe_to.topic_filter),
+                            subscribe_to.qos,
+                        );
+                    }
                     SubscriptionUpdate::Unsubscribe(unsubscribe_from) => {
-                        target_subscriptions.remove(&*unsubscribe_from)
+                        if target_subscriptions.remove(&*unsubscribe_from).is_none() {
+                            target_unsubscriptions.insert(unsubscribe_from, true);
+                        }
                     }
                 };
             }
 
             let mut pending_subscriptions: std::collections::VecDeque<_> = Default::default();
             for (topic_filter, &qos) in &target_subscriptions {
-                if current_subscriptions.get(topic_filter) != Some(&qos) {
-                    // Current subscription doesn't exist, or exists but has different QoS
-                    pending_subscriptions.push_back(crate::proto::SubscribeTo {
-                        topic_filter: topic_filter.clone().into_owned(),
-                        qos,
-                    });
-                }
+                pending_subscriptions.push_back(crate::proto::SubscribeTo {
+                    topic_filter: topic_filter.clone().into_owned(),
+                    qos,
+                });
             }
 
             let mut pending_unsubscriptions: std::collections::VecDeque<_> = Default::default();
-            for topic_filter in current_subscriptions.keys() {
-                if !target_subscriptions.contains_key(topic_filter) {
-                    pending_unsubscriptions.push_back(topic_filter.clone().into_owned());
```

Comment:
I don't really understand this original logic here. It is saying that for a subscription waiting for an ack to come back, if it does not have a matching subscription waiting to get sent, then it will try to unsubscribe?